bentoml/_internal/bento/docker.py

```diff
@@ -1,158 +1,157 @@
-
-        Example results:
-
-        * bentoml/bento-server:`devel-python3.7-debian`
-        * bentoml/bento-server:`0.14.0-python3.8-ubi8-cudnn`
-        * bentoml/bento-server:`0.14.0-python3.7-ami2-runtime`
-
+CUDA_SUPPORTED_DISTRO = [
+    v for v in DOCKER_SUPPORTED_DISTRO if "cuda" in DOCKER_SUPPORTED_DISTRO[v]
+]
+CONDA_SUPPORTED_DISTRO = ["debian-miniconda", "alpine-miniconda"]
 
-    Example::
 
-    from bento import svc
+@attr.define(slots=True, frozen=True)
+class DistroSpec:
+    name: str = attr.field()
+    image: str = attr.field(kw_only=True)
+    release_type: str = attr.field(kw_only=True)
 
-    svc.build(
-        ...
-        env=dict(
-            docker=dict(
-                distro='debian',
-                python='3.8',
-                gpu=True
-            )
-        )
+    python_version: t.List[str] = attr.field(
+        validator=attr.validators.deep_iterable(
+            lambda _, __, value: value in SUPPORTED_PYTHON_VERSION,
+            iterable_validator=attr.validators.instance_of(list),
+        ),
     )
-    """
-
-    def __init__(
-        self,
-        distro: str,
-        python_version: t.Optional[str] = None,
-        gpu: t.Optional[bool] = None,
-        devel: t.Optional[bool] = None,
-    ) -> None:
-        if gpu is None:
-            gpu = False
-
-        if devel:
-            # Use `devel` nightly built base image
-            if gpu:
-                raise BentoMLException("`devel` base image does not support GPU")
-            if distro not in SUPPORTED_DEVEL_DISTROS:
-                raise BentoMLException(
-                    f"`devel` base image does not support distro=`{distro}`"
-                )
-
-            self._release_type: str = "devel"
-        else:
-            self._release_type = "cudnn" if gpu else "runtime"
-
-        if gpu and distro not in SUPPORTED_GPU_DISTROS:
-            raise BentoMLException(
-                f"distro '{distro}' with GPU={gpu} is not supported. "
-                f"GPU-supported distros are: {SUPPORTED_GPU_DISTROS} "
-            )
 
-        if distro not in SUPPORTED_RELEASES_COMBINATION[self._release_type]:
-            raise BentoMLException(
-                f"{distro} is not yet supported. "
-                "Supported distros: "
-                f"{SUPPORTED_RELEASES_COMBINATION[self._release_type]}"
+    cuda_version: t.List[str] = attr.field(
+        default=None,
+        validator=attr.validators.optional(
+            attr.validators.deep_iterable(
+                lambda _, __, value: value in SUPPORTED_CUDA_VERSION.values(),
             )
+        ),
+    )
 
-        self._distro: str = distro
-
-        if python_version:
-            assert re.match(r"^[2,3]\.[0-9]{1,2}$", python_version), (
-                f"python_version '{python_version}' should be in format "
-                "'{major}.{minor}', e.g. 'python_version=\"3.7\"'"
+    architecture: t.Optional[t.List[str]] = attr.field(
```

Comment:
should we have this option here? seems better to provide it only during `bentoml containerize`

Reply:
this is the architecture that is supported per distro. The `platform` args is irrelevant here.